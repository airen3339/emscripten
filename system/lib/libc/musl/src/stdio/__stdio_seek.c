#include "stdio_impl.h"

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
	off_t ret;
#ifdef SYS__llseek
	off_t off_hi =
#ifndef __EMSCRIPTEN__
		off >> 32;
#else
		0;
#endif
	if (syscall(SYS__llseek, f->fd, off_hi, off, &ret, whence)<0)
		ret = -1;
#else
	ret = syscall(SYS_lseek, f->fd, off, whence);
#endif
	return ret;
}
