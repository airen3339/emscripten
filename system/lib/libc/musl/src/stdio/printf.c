#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int iprintf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfiprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vfprintf(FILE *restrict f, const char *restrict fmt, va_list ap);

int __small_printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

