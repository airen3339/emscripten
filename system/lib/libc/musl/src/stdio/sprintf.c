#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int siprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsiprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vsprintf(char *restrict s, const char *restrict fmt, va_list ap);

int __small_sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

