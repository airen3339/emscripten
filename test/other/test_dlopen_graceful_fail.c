// Emscripten's stub implementation of dlopen() hard aborts execution.
// Some users want to instead have dlopen() perform a graceful "return 0" failure
// where they handle the dlopen failures in code, so they want to override
// Emscripten's built-in JS library functions.

// This test verifies that the ability to override dlopen() to not hard abort
// will not regress.
#include <dlfcn.h>
#include <stdio.h>

int main()
{
  printf("Attempting to dlopen\n");
  void *dll = dlopen("nonexistent.dll", 0);
  printf("dlopen completed: ptr %p. error: %s\n", dll, dlerror());
  dlclose(0);
  printf("dlclose(0) completed.\n");
  if (!dll) printf("dlopen failed gracefully!\n");
  return 0;
}
