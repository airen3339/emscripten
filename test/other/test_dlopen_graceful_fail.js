#if !MAIN_MODULE
#if !SIDE_MODULE
// This code represents stable end-user facing API. Changing any part of this file
// will break end users!
mergeInto(LibraryManager.library, {
  _dlopen_js__deps: ['$warnOnce'],
  _dlopen_js: function(handle) {
#if ASSERTIONS
    warnOnce('_dlopen_js: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.');
#endif
  },

  _emscripten_dlopen_js__deps: ['$warnOnce'],
  _emscripten_dlopen_js: function(filename, flags, user_data, onsuccess, onerror) {
#if ASSERTIONS
    warnOnce('_emscripten_dlopen_js: Unable to open DLL ' + UTF8ToString(filename) + '! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.');
#endif
  },

  _dlsym_js__deps: ['$warnOnce'],
  _dlsym_js: function(handle, symbol, symbolIndex) {
#if ASSERTIONS
    warnOnce('_dlsym_js: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.');
#endif
  },

  _dlsym_catchup_js__deps: ['$warnOnce'],
  _dlsym_catchup_js: function(handle, symbolIndex) {
#if ASSERTIONS
    warnOnce('_dlsym_catchup_js: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.');
#endif
  },

  dlopen__deps: ['$warnOnce'],
  dlopen: function(handle) {
#if ASSERTIONS
    warnOnce('dlopen: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.');
#endif
  },

  emscripten_dlopen__deps: ['$warnOnce'],
  emscripten_dlopen: function(handle, onsuccess, onerror, user_data) {
#if ASSERTIONS
    warnOnce('emscripten_dlopen: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.');
#endif
  },

  __dlsym__deps: ['$warnOnce'],
  __dlsym: function(handle, symbol) {
#if ASSERTIONS
    warnOnce('__dlsym: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.');
#endif
  },
});
#endif
#endif
