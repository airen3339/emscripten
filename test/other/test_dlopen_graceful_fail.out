Attempting to dlopen
warning: dlopen: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.
dlopen completed: ptr 0. error: (null)
dlclose(0) completed.
dlopen failed gracefully!